File: tests/instructions/cancel_order.ts
import { ASSOCIATED_TOKEN_PROGRAM_ID, TOKEN_PROGRAM_ID } from '@solana/spl-token';
import { PublicKey, SystemProgram, SYSVAR_RENT_PUBKEY, TransactionInstruction } from '@solana/web3.js';
import BN from 'bn.js';
import * as borsh from 'borsh';
import { Assignable, LimitOrderInstruction } from '../instruction';

class CancelOrder extends Assignable {
    toBuffer() {
      return Buffer.from(borsh.serialize(CancelOrderSchema, this));
    }
  }
  
  const CancelOrderSchema = new Map([
    [
      CancelOrder,
      {
        kind: 'struct',
        fields: [
          ['instruction', 'u8'],
          ['id' , 'u64'],
          ['amount', 'u64']
        ],
      },
    ],
  ]);
  
  export function buildCancelOrder(props : {
    id : BN,
    amount : BN,
    user : PublicKey;
    btc_order_book : PublicKey;
    order_book_admin_pubkey : PublicKey;
    token_mint_a : PublicKey;
    user_token_account_a : PublicKey;
    mediator_vault : PublicKey;
    program_id : PublicKey;
  }) {
    const cancel_ix = new CancelOrder({
        instruction : LimitOrderInstruction.CancelOrder,
        id : props.id,
        amount : props.amount
    });
  
    return new TransactionInstruction({
        keys : [
          {
              pubkey : props.user,
              isSigner : false,
              isWritable : true
          },
            {
                pubkey : props.btc_order_book,
                isSigner : false,
                isWritable : true
            },
            {
              pubkey : props.order_book_admin_pubkey,
              isSigner : false,
              isWritable : true
          },
            {
              pubkey : props.token_mint_a,
              isSigner : false,
              isWritable : false
          },
          {
            pubkey : props.user_token_account_a,
            isSigner : false,
            isWritable : true
        },
          {
            pubkey : props.mediator_vault,
            isSigner : false,
            isWritable : true
           },
          {
            pubkey : TOKEN_PROGRAM_ID,
            isSigner : false,
            isWritable : false
          },
          {
              pubkey : ASSOCIATED_TOKEN_PROGRAM_ID,
              isSigner : false,
              isWritable : false
          },
          {
            pubkey : SystemProgram.programId,
            isSigner : false,
            isWritable : false
        },
        ],
        programId : props.program_id,
        data : cancel_ix.toBuffer()
    })
  }

File: src/instructions/cancel_order.rs
use crate::{error::ApplicationError, state::Manager};
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{
    account_info::AccountInfo, entrypoint::ProgramResult, program::invoke_signed,
    program_error::ProgramError, pubkey::Pubkey,
};
use spl_associated_token_account::instruction as associated_token_account_instruction;
use spl_token::{instruction as token_instruction, state::Account as TokenAccount};

#[derive(BorshSerialize, BorshDeserialize, Debug)]
pub struct CancelOrder {
    id : u64,
    amount: u64,
}

impl CancelOrder {
    pub fn cancel_order(
        program_id: &Pubkey,
        accounts: &[AccountInfo],
        args: CancelOrder,
    ) -> ProgramResult {
        
        let [
            user, // user that's create the order
            manager, //manager config_account
            manager_auth, // manager auth
            token_mint_a, // token_mint that user want to trade for
            user_token_account_a, // user token_account for mint_a
            mediator_vault, // vault where user token are stored
            token_program_id,
            associated_token_program,
            system_program
        ] =
            accounts
        else {
            return Err(ProgramError::NotEnoughAccountKeys);
        };

        let manager_account_data = Manager::try_from_slice(&manager.data.borrow()[..])?;

        let manager_account_seeds = &[
            b"btc_order_book",
            manager_auth.key.as_ref(),
            &[manager_account_data.bump],
        ];

        let order_book_key = Pubkey::create_program_address(manager_account_seeds, program_id)?;

        if order_book_key != *manager.key {
            return Err(ApplicationError::MismatchOrderbookKey.into());
        }

        // send user back its fund from mediator_vault
        invoke_signed(
            &token_instruction::transfer(
                token_program_id.key,
                mediator_vault.key,
                user_token_account_a.key,
                manager.key,
                &[],
                args.amount,
            )?,
            &[
                user.clone(),
                token_mint_a.clone(),
                mediator_vault.clone(),
                user_token_account_a.clone(),
                manager_auth.clone(),
                manager.clone(),
                token_program_id.clone(),
            ],
            &[manager_account_seeds],
        )?;

        Ok(())
    }
}
